### DISCUSSION TITLE
How to get Windows 10 to recognize my  USB HID deivce?

### DISCUSSION THREAD

**Discussion by xayhewalo** [hathach/tinyusb#1594] [Q&A]
## Problem
I've successfully built and flashed a SAMD11 XPlained Pro with the [HID composite](https://github.com/hathach/tinyusb/tree/master/examples/device/hid_composite) example, but Windows 10 cannot recognize the device when I connect the board via the target USB port. The onboard LED is flashing so that leads me to believe I've configured everything correctly.

![unknown_usb](https://user-images.githubusercontent.com/7256157/183445551-04c6ca8d-d24a-42fe-9b23-c89c6bac5ed4.JPG)
![unknown_usb1](https://user-images.githubusercontent.com/7256157/183446023-7351b9a1-4a6b-4ffd-a61e-77eb995b54b0.JPG)

My ultimate goal is to get make a USB HID gamepad, and the software I'm using to test the gamepad requires the host recognize the device as so. I believe this error occurs because [`tud_hid_get_report_cb`](https://github.com/hathach/tinyusb/blob/251ce9899039fbdf8e749342bc6504e7438d2e88/examples/device/hid_composite/src/main.c#L244) is not implemented. I should mention that I'm a USB beginner so I may be making some incorrect assumptions. Please correct me if I'm wrong.

I've read through the [USB HID specification](https://www.usb.org/sites/default/files/hid1_11.pdf) and I believe `tud_hid_get_report_cb` is supposed to implement the `Get_Report Request` in section 7.2.1 on page 49 on the HID document. I'm not certain on this assumption because the comment for `tud_hid_get_report_cb` says the function should return the length of the request.  However, the HID document refers to section 9.4  of the USB 2.0 specification (`Standard Device Requests`), which states requests should contain `bmRequestType`, `bRequest`, `wValue`, `wIndex`, `wLength`, and `Data`. I believe I'm misunderstanding something, so I'm asking this community for guidance. I tried just returning `reqlen` from `tud_hid_get_report_cb`, but that didn' resolve the issue.

## Toolchain Details
I'm using Microchip Studio on Windows 10. I've configured the clock of the USB and nothing else using Atmel Start.
![unknown_usb2](https://user-images.githubusercontent.com/7256157/183453310-b3976ae6-9223-4475-a028-3fa14fd16423.JPG)
I added `../tinyusb/src`, `../tinusb/hw`, and `../tinyusb/hw/bsp/samd11/boards/samd11_xplained` to Microchip Studio's include paths. I added all the *.c source files in [`rules.mk`](https://github.com/hathach/tinyusb/blob/251ce9899039fbdf8e749342bc6504e7438d2e88/examples/rules.mk#L26) to the Microchip Studio project (I don't think I needed them all, but *shurg* I didn't want to miss anything). I added [`family.c`](https://github.com/hathach/tinyusb/blob/master/hw/bsp/samd11/family.c) from `\tinyusb\hw\bsp\samd11\` to the Microchip Studio project too.

I copy and pasted [`tusb_config.h`](https://github.com/hathach/tinyusb/blob/master/examples/device/hid_composite/src/tusb_config.h) from the HID composite example and defined `CFG_TUSB_MCU` and `CFG_TUSB_RHPORT0_MODE` within the file.
```
...
// defined by compiler flags for flexibility
#ifndef CFG_TUSB_MCU
#define CFG_TUSB_MCU OPT_MCU_SAMD11
#endif

#define CFG_TUSB_RHPORT0_MODE OPT_MODE_DEVICE
...
```
I also copy and pasted [`usb_descriptors.c`](https://github.com/hathach/tinyusb/blob/master/examples/device/hid_composite/src/usb_descriptors.c) and it's header file into the Microchip Studio project.
In `family.c` I commented out `_init` as it's already defined in Atmel's driver and TinyUSB's implementation doesn't seem like it's doing anything. I also commented out `USB_Handler` in Atmel's driver, `\hpl\usb\hpl_usb.c`, in favor of TinyUSB's implementation. Was this the correct thing to do?

**Comment by xayhewalo** [comment]
Anyone?

I can run the [board test](https://github.com/hathach/tinyusb/tree/master/examples/device/board_test) example after doing similar steps that I outlined above. From [@2400audio's post](https://github.com/hathach/tinyusb/discussions/1069) post and from [@zlittell reply](https://github.com/hathach/tinyusb/discussions/633#discussioncomment-341521) other people have integrate TinyUSB into MicroChip Studio. Can anyone point a minimal example where a Windows recognized a TinyUSB device?

**Reply by hathach** [reply]
board_test only test uart/led/button and does not test usb at all. You should try any of other examples , simplest is probably dfu_runtime. Have you tried compiling with the stock make to rule out hw issue.

**Reply by xayhewalo** [reply]
The hardware is working because the usb keyboard and mouse examples from Atmel are properly recognized by Windows. Same board, same usb cable, same usb port. I run into issues when using tinyusb.

I tried the dfu runtime example and I'm getting the same issue. This is what I did:
- Configured the usb clock speed exactly as in the original post
- Cloned tinyusb into the root directory of the Atmel Start project
- Copy and pasted [`usb_descriptors.c`](https://github.com/hathach/tinyusb/blob/master/examples/device/dfu_runtime/src/usb_descriptors.c)
- Modified line 35 [`tusb_config.h`](https://github.com/hathach/tinyusb/blob/master/examples/device/dfu_runtime/src/tusb_config.h) like so and copy and pasted it into the root directory:
```
- #error CFG_TUSB_MCU must be defined
+ #define  CFG_TUSB_MCU OPT_MCU_SAMD11
```
- Included `tusb_config.h` and `<atmel_start.h>` in [`main.c`](https://github.com/hathach/tinyusb/blob/master/examples/device/dfu_runtime/src/main.c) and insert `atmel_start_init()` as the first line of the `main` function
- Defined `CFG_TUSB_RHPORT0_MODE` to `OPT_MODE_DEVICE` in `tinyusb/src/tusb_option.h`
- Added `tinyusb\src\tusb.c` to the Microchip Studio project so the IDE could find function definitions
- Added `tinyusb\src\device\usbd.c`
- Added `tinyusb\src\device\usbd_control.c` to the project
- Added `tinyusb\src\portable\microchip\samd\dcd_samd.c` to the project
- Added `tinyusb\src\class\dfu\dfu_device.c` and `tinyusb\src\class\dfu\dfu_rt_device.c`
- Added `tinyusb\hw\bsp\samd11\family.c` to the project
- Modified line 29 of `family.c` like so
```
- #include "board.h"
+ #include "bsp/samd11/boards/samd11_xplained/board.h"
```
- Commented out `_init()` at the bottom of `family.c`
- Commented out `USB_Handler()` in Atmel's driver `hpl\usb\hpl_usb.c`
- Added `tinyusb/src/` and `tinyusb/hw` to the include paths

Does anyone see any mistakes in the steps I took above? This was enough to get the code to compile and blink the LED on the board. I moved to a Linux platform to try to run `dfu-util` but the host didn't find any DFU devices when I plugged in the samd11 board. `lsusb` also did not detect the board. Any help would be greatly appreciated.

**Reply by hathach** [reply]
I don't use Atmel IDE to comment on the setup, this is probably an IDE integration issue. You tried compiling with the gcc make and compare with IDE compile command/symbol/output etc ...

**Reply by xayhewalo** [reply]
Are there anymore macros or functions I need to define to get tinyusb running?

I'll try compiling with a make file and report back. Likely will switch to a Linux platform instead of WSL.

**Reply by xayhewalo** [reply]
Still troubleshooting, but wanted to add that I'm using ARM/GNU C Compiler 6.3.1, ARM/GNU Linker 6.3.1, and CMSIS 5.4.0. All these versions are compatible right?

**Reply by hathach** [reply]
There is no issue with gcc6 that I could remember.
